src/golang/cmd/server/handler/export_function.go

```diff
@@ -125,13 +135,92 @@ func (h *ExportFunctionHandler) Perform(ctx context.Context, interfaceArgs inter
 		return emptyResp, http.StatusInternalServerError, errors.Wrap(err, "Unable to get function from storage")
 	}
 
+	readableProgram, err := extractUserReadableCode(program, operatorObject.Name)
+	if err != nil {
+		return emptyResp, http.StatusInternalServerError, errors.Wrap(err, "Unable to export function code")
+	}
+
 	return &exportFunctionResponse{
-		fileName: args.operatorId.String(),
-		program:  bytes.NewBuffer(program),
+		fileName: operatorObject.Name,
+		program:  bytes.NewBuffer(readableProgram),
 	}, http.StatusOK, nil
 }
 
 func (*ExportFunctionHandler) SendResponse(w http.ResponseWriter, interfaceResp interface{}) {
 	resp := interfaceResp.(*exportFunctionResponse)
 	response.SendSmallFileResponse(w, resp.fileName, resp.program)
 }
+
+// extractUserReadableCode takes the zipped function code and only returns a zipped file
+// containing the human-readable parts, i.e. source code, requirements.txt, and python_version.txt
+// If no source code is found, it simply returns the original contents.
+// In all cases, it renames the top-level directory to `operatorName`.
+func extractUserReadableCode(data []byte, operatorName string) ([]byte, error) {
+	zipReader, err := zip.NewReader(bytes.NewReader(data), int64(len(data)))
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if there is a source file, since older SDK clients did not generate this file
+	hasSourceFile := false
+	for _, zipFile := range zipReader.File {
+		sourceFileName := fmt.Sprintf("%s.py", operatorName)
+		parts := strings.Split(zipFile.Name, "/")
+		if len(parts) == 2 && parts[1] == sourceFileName {
+			hasSourceFile = true
+			break
+		}
+	}
+
+	buf := new(bytes.Buffer) // This is where the new zipped file is written to
+	zipWriter := zip.NewWriter(buf)
+
+	for _, zipFile := range zipReader.File {
+		parts := strings.Split(zipFile.Name, "/")
+		if hasSourceFile &&
+			len(parts) == 2 &&
+			(parts[1] == modelFile || parts[1] == modelPickleFile) {
+			// There is a source file so we can skip the files that are not user-friendly to read
+			continue
+		}
+
+		// Generate a new file name using `operatorName`, because it is more user-friendly to read
+		// than the current directory name that is a unique UUID
+		parts[0] = operatorName
+		zipFileName := strings.Join(parts, "/")
+
+		if err := writeZipFile(zipWriter, zipFile, zipFileName); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := zipWriter.Close(); err != nil {
```

Comment:
I'm not familiar with the zipWriter interface, is there a reason we don't `defer` this after the zipWriter creation?

Reply:
We have to explicitly close the `zipWriter` before we can read its content: https://stackoverflow.com/a/55623815